src/main/java/com/microsoft/graph/authentication/TokenCredentialAuthProvider.java

```diff
@@ -0,0 +1,91 @@
+
+public class TokenCredentialAuthProvider implements ICoreAuthenticationProvider , IAuthenticationProvider {
+
+    //TokenCredential expected from user
+    private TokenCredential tokenCredential;
+    //Context options which can be optionally set by the user
+    private TokenRequestContext context;
+
+    
+    //TODO: Upon further review from Peter, this should include a Null check in the case that this is the first request
+    //There should be just one constructor as scopes should probably not be optional
+    /**
+     * Creates an Authentication provider using a passed in TokenCredential
+     *
+     * @param tokenCredential Credential object inheriting the TokenCredential interface used to instantiate the Auth Provider
+     * @throws AuthenticationException exception occurs if the TokenCredential parameter is null
+     */
+    public  TokenCredentialAuthProvider(TokenCredential tokenCredential) throws AuthenticationException {
+        if(tokenCredential == null) {
+            throw new AuthenticationException(new Error(Codes.InvalidArgument,
+                    String.format(Messages.NullParameter, "TokenCredential"))
+                    ,new IllegalArgumentException());
+        }
+
+        this.tokenCredential = tokenCredential;
+        this.context = new TokenRequestContext();
+    }
+
+    /**
+     * Creates an Authentication provider using a TokenCredential and list of scopes
+     *
+     * @param tokenCredential Credential object inheriting the TokenCredential interface used to instantiate the Auth Provider
+     * @param scopes Specified desired scopes of the Auth Provider
+     * @throws AuthenticationException exception occurs if the TokenCredential parameter is null
+     */
+    public TokenCredentialAuthProvider(TokenCredential tokenCredential, List<String> scopes) throws AuthenticationException {
+        this(tokenCredential);
+        this.context.setScopes(scopes);
+    }
+
+    /**
+     * Authenticates the request
+     *
+     * @param request the request to authenticate
+     */
+    @Override
+    public void authenticateRequest(IHttpRequest request) {
+        request.addHeader(AuthConstants.AUTHORIZATION_HEADER, AuthConstants.BEARER + getAccessToken());
+    }
+
+    /**
+     * Authenticates the request
+     *
+     * @param request the request to authenticate
+     * @return Request with Authorization header added to it
+     */
+    @Override
+    public Request authenticateRequest(Request request) {
+        return request.newBuilder()
+                    .addHeader(AuthConstants.AUTHORIZATION_HEADER, AuthConstants.BEARER + getAccessToken())
+                    .build();
+    }
+
+    /**
+     * Returns an AccessToken as a string
+     *
+     * @return String representing the retrieved AccessToken
+     */
+    String getAccessToken() {
+        String accessToken = null;
+        try {
+            AccessToken token = this.tokenCredential.getToken(this.context).block();
```

Comment:
To be addressed another day, but this will block indefinitely. We'll also want to handle errors raised from getToken

Reply:
@MIchaelMainer I just addressed this with the following commit, can you review and let me know whether this meets your expectations? https://github.com/microsoftgraph/msgraph-sdk-java-core/pull/42/commits/4622566c4c7c61d5591d12893fac21d8988d3e59